Add a url keyword or argument to Agent#initialize

CharlieWWW94 opened the pull request:
Adds the url: kwarg to the Ronin::Web::Browers::Agent initializer, as outlined in issue #16. Upon initialization the browser will navigate to this url.

Hope this contribution is useful - did my best to conform to code conventions across the rest of the repo. Please do let me know of any changes that are needed. Eager to contribute in any way that might be helpful 👍

postmodern commented on lib/ronin/web/browser/agent.rb:
```diff
@@ -39,7 +39,7 @@ class Agent < Ferrum::Browser
 
         #
-        # Initializes the browser agent.
+        # Initializes the browser agent and navigates to the url, if provided.
```
I would add ", and optionally navigates" since the `url:` keyword isn't required.

postmodern commented on lib/ronin/web/browser/agent.rb:
```diff
@@ -79,10 +83,17 @@ def initialize(visible:  false,
           @proxy    = proxy
 
           super(headless: headless, proxy: proxy, **kwargs)
+          go_to url if url
```
I would put parenthesis around `url` to avoid ambiguity with the `if`.

postmodern reviewed (changes requested):
Minor stylistic changes.

I would also copy/paste the `@param` tag into `lib/ronin/web/browser.rb` for the `.new` method.

Also, if it's not too much work, could you rebase against the `0.2.0` branch? I like to merge new features into the next upcoming minor version branch. Otherwise I can manually merge your branch into `0.2.0`.

CharlieWWW94 commented:
> Minor stylistic changes.
> 
> I would also copy/paste the `@param` tag into `lib/ronin/web/browser.rb` for the `.new` method.
> 
> Also, if it's not too much work, could you rebase against the `0.2.0` branch? I like to merge new features into the next upcoming minor version branch. Otherwise I can manually merge your branch into `0.2.0`.

Thanks for the feedback! Have completed those, rebased against 0.2.0 and updated the base branch of the PR to 0.2.0.

postmodern commented on lib/ronin/web/browser/agent.rb:
```diff
@@ -79,10 +83,17 @@ def initialize(visible:  false,
+                when nil, String then url
+                else
+                  raise(ArgumentError,"invalid url value (#{url.inspect}), must be either String, or nil")
+                end
```
Is this code really necessary? We could let `go_to()` just try to handle the `url` value if it's non-`nil`.

If it is needed, you could move the `case` statement down below around the `go_to`.

```ruby
case url
when String then go_to(url)
when nil    then # no-op
else
  ...
end
```

CharlieWWW94 replied:
Have refactored with your revised case statement rather than removed - the error messaging does seem much nicer/clearer from a user perspective when raising the ArgumentError, rather than allowing Ferrum to handle it. Just let me know if you disagree and can remove.

postmodern commented on lib/ronin/web/browser.rb:
```diff
@@ -117,6 +117,9 @@ module Browser
       # @option kwargs [Boolean] headless (true)
       #   Controls whether the browser will start in headless or visible mode.
       #
+      # @option kwargs [String, nil] url (nil)
```
`(nil)` can be omitted.

postmodern reviewed (changes requested):
Minor changes. Looks good except for those two things.

postmodern commented on lib/ronin/web/browser/agent.rb:
```diff
@@ -83,6 +87,13 @@ def initialize(visible:  false,
+          when nil # no-op
+          else
+            raise(ArgumentError,"invalid url value (#{url.inspect}), must be either String, or nil")
+          end
```
OK you're going to hate me, but I did some testing. Appears that `go_to` uses `Addressable::URI` to parse/coerce URLs, so it accepts Strings, `URI`s, and anything that responds to `to_str`. I think a simple `if url` check is good enough. Last comment I swear! :)

CharlieWWW94 replied:
Haha! All good, will sort now.

postmodern reviewed (changes requested):
Last request for change I swear! Did some testing to figure out what values `go_to` actually accepts. Turns out `URI::HTTP` and anything that `Addressable::URI` can accept are valid.
